cirq-core/cirq/study/resolver.py

```diff
@@ -179,7 +180,11 @@ def value_of(
         if not recursive:
             # Resolves one step at a time. For example:
             # a.subs({a: b, b: c}) == b
-            v = value.subs(self.param_dict, simultaneous=True)
+            try:
+                v = value.subs(self.param_dict, simultaneous=True)
+            except sympy.SympifyError:
```

Comment:
What do you mean by "not parsable"? An actual syntax error? If so, then this is probably a user mistake and it's better to let the exception propagate out of here. Then the PR could catch it in resolver_test.py rather than in resolver.py. WDYT?

Reply:
The actual sympy bug is here:
https://github.com/sympy/sympy/pull/23887

Their definition is "sympifiable" whatever that means.  I think that it is either a sympy object or a basic python type, possibly a few other cases.  I think it is talking about objects foreign to the sympy library.  In our case, that is the symbol 'NotImplemented'.

I am hesitant to change the existing behavior.  The test currently asserts that resolving {'b': NotImplemented} would just leave sympy.Symbol('b') alone.  I am loathe to change this behavior right now, since there is a lot of special handling around NotImplemented.

Reply:
I agree we should preserve existing behavior.

I have two conerns:
* the fact that if one of the values is unsympifiable then none of the substitutions are made,
* the fact that we're failing to communicate a potential mistake to the user as soon as we discover it (we could be passing an unsympifiable object as a result of a simple typo in the initialization of the ParamDict).

Is `NotImplemented` the only offender? If so, could we just filter them out by saying something like

```python
v = value.subs({k: w for k, w in self.param_dict.values() if w is not NotImplemented}, simultaneous=True)
```
?

Reply:
I changed this to raise a ValueError instead, explaining what the problem substitution is.

We cannot do your suggestion, because the test case is a function that defines _resolved_value_ as `NotImplemented`.  This is indistinguishable from something like `Sympy.Symbol('b')+Sympy.Symbol('c')` which also has an unimplemented _resolved_value_.   

This does change the behavior of the function slightly, but, as you pointed out, the previous PR would also change the value as some additional variables would not be substituted.

I think this is the best solution we can do, since this preserves the behavior 'the variable could not be substituted' but makes it explicit as a ValueError rather than silently refusing to substitute it.  It's also quite a buried case, because it involves the user specifying something that was unresolvable in the first place.